test/framework/src/main/java/org/elasticsearch/test/AbstractQueryTestCase.java

```diff
@@ -529,8 +532,47 @@ public void testToQuery() throws IOException {
         }
     }
 
-    protected QueryBuilder rewriteQuery(QB queryBuilder, QueryRewriteContext rewriteContext) throws IOException {
-        QueryBuilder rewritten = rewriteAndFetch(queryBuilder, rewriteContext);
+    /**
+     * Simulate rewriting the query builder exclusively on the data node.
+     * <br/>
+     * <br/>
+     * NOTE: This simulation does not reflect how the query builder will be rewritten in production.
+     * See {@link AbstractQueryTestCase#rewriteQuery(AbstractQueryBuilder, QueryRewriteContext, SearchExecutionContext)} for a more accurate
+     * simulation.
+     *
+     * @param queryBuilder The query builder to rewrite
+     * @param shardRewriteContext The data node rewrite context
+     * @return The rewritten query builder
+     * @throws IOException
+     */
+    protected QueryBuilder rewriteQuery(QB queryBuilder, SearchExecutionContext shardRewriteContext) throws IOException {
+        QueryBuilder rewritten = rewriteAndFetch(queryBuilder, shardRewriteContext);
+        // extra safety to fail fast - serialize the rewritten version to ensure it's serializable.
+        assertSerialization(rewritten);
+        return rewritten;
+    }
+
+    /**
+     * Simulate rewriting the query builder in stages across the coordinator node and data node.
+     * It is rewritten on the coordinator node first, then again on the data node.
+     *
+     * @param queryBuilder The query builder to rewrite
+     * @param coordinatorRewriteContext the coordinator node rewrite context
+     * @param shardRewriteContext The data node rewrite context
+     * @return The rewritten query builder
+     * @throws IOException
+     */
+    protected QueryBuilder rewriteQuery(
+        QB queryBuilder,
+        QueryRewriteContext coordinatorRewriteContext,
+        SearchExecutionContext shardRewriteContext
+    ) throws IOException {
+        // The first rewriteAndFetch call simulates rewriting on the coordinator node
+        // The second rewriteAndFetch call simulates rewriting on the shard
+        QueryBuilder rewritten = rewriteAndFetch(queryBuilder, coordinatorRewriteContext);
+        // extra safety to fail fast - serialize the rewritten version to ensure it's serializable.
+        assertSerialization(rewritten);
```

Comment:
This checks that the query can be serialized after being rewritten on the coordinator node, which is an important check that would otherwise require setting up a multi-node integration test environment.